src/Microsoft.OData.Edm/Csdl/CsdlWriter.cs

```diff
@@ -132,6 +208,15 @@ protected virtual void WriteCsdl()
             // nothing here
         }
 
+        /// <summary>
+        /// Asynchronously Writes CSDL output.
+        /// </summary>
+        /// <returns>Task represents an asynchronous operation that may or may not return a result.</returns>
+        protected virtual Task WriteCsdlAsync()

```

Comment:
This is an observation, not a suggestion for change, but I find this virtual `Write` methods that do nothing to be a potential source for bugs. The fact that it has a default implementation means that child classes are not required to implement it, but a `Write` method that does nothing is in my opinion unintuitive and unexpected. So, there's no practical use case of child class that does not override this method. That makes this default implementation practically useless. I think this would be better off being an abstract method than empty virtual method. That way subclasses would be forced to provide an implementation, and if they actually wanted a no-op, they would have to provide a no-op implementation explicitly. This refers to the existing `WriteCsdl` method. This would be a breaking change, but I think the impact would small because it's not common for services to create custom subclasses of this class. And since ODL 8 is major version release, it would be a good opportunity for such change.

This is not just a hypothetical problem. We have actually gotten bitten by something similar. We had a custom `TextWriter` implementation overrode `TextWriter.Writer(char[])` method but not the `TextWriter.Write(char)` overload. It just so happens that the `TextWriter.Writer(char)` default implementation is a no-op (I would expect it to call `TextWrtier.Write(new [] {char})` at least). But since the compiler didn't force us to implement this method, and we didn't have test coverage for this, we did not catch the bug when the PR was merged. Luckily the issue was caught later on and a patch made.

Reply:
@habbes Thanks for bringing this up. **I see a lot of similar methods which are virtual with empty implementation across the project**. In this case, `base class` expects the` child classes` will `override `these methods and provide their own implementation. To mark these methods as `abstract`, we will need also to mark the `base classes` as `abstract`.

Reply:
My point was that an empty virtual method that's always expected to write something is a bad idea. I would prefer that they bet turned abstract instead, but I don't know if there are unknown repercussions that would follow, and maybe that could derail this PR.

Reply:
On a side note, interesting discussion about `TextWriter.Write(char)` being an empty virtual method instead of an abstract method: https://github.com/dotnet/runtime/discussions/104345

Reply:
I also wanted to point out that this is not a blocker for me.

Reply:
I will close this thread as resolved